cpp/include/cuvs/neighbors/cagra_optimize.hpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/mdspan_types.hpp>
+#include <raft/core/resources.hpp>
+
+namespace cuvs::neighbors::cagra::helpers {
+
+/**
+ * @brief Optimize a KNN graph into a CAGRA graph.
+ *
+ * This function prunes and optimizes a k-NN graph to create a more efficient
+ * CAGRA search graph. The input graph must be on host memory as the optimization
+ * algorithm uses CPU-based processing.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *   raft::resources res;
+ *
+ *   // If starting with device data, copy to host first:
+ *   auto d_knn = raft::make_device_matrix<uint32_t, int64_t>(res, N, K_in);
```

Comment:
I think we can exclude this from the example- we don't need to assume user has the knn on device.